### DISCUSSION TITLE
Managing different thrust maps according to engine ratings

### DISCUSSION THREAD

**Discussion by agodemar** [JSBSim-Team/jsbsim#389] [Support requests]
Referring to the discussion #381 on "Accurate TSFC - Thrust Specific Fuel Consumption", the following is a request regarding thrust modelling.

I have in my model a selector function that picks one of the available [engine ratings](https://en.wikibooks.org/wiki/Jet_Propulsion/Engine_ratings) (i.e. engine states). Ratings range from Maximum Takeoff thrust (MTO), to Maximum Climb thrust (MCL), Cruise, Flight Idle, etc. For each rating I have tables that map thrust _T_ versus Mach number _M_, altitude (ASL) _h_, and ISA Delta-Temperature _ISA-DT_. Similarly, I have maps of TSFC as functions of the same variables. 
I have such models for both turbofan engines and turboprop engines.

How can this be modelled in JSBSim? 

I can have a config file for a `<turbine_engine>` where I can define a `<function name="MilThrust">` or a `<function name="IdleThrust">`. But how all available ratings can be included in a model?

**Comment by seanmcleod** [comment]
Use a `fcs/engine-rating-value` property as a table breakpoint?

```xml
<table>
   <independentVar lookup="row">    fcs/row-value    </independentVar>
   <independentVar lookup="column"> fcs/column-value </independentVar>
   <independentVar lookup="table">  fcs/engine-rating-value  </independentVar>

   <tableData breakPoint="1">
           -1.0    1.0
      0.0  1.0000  2.0000
      1.0  3.0000  4.0000
   </tableData>

   <tableData breakPoint="2">
           0.0     10.0
      2.0  1.0000  2.0000
      3.0  3.0000  4.0000
   </tableData>

   <tableData breakPoint="3">
            0.0     10.0    20.0
       2.0  1.0000  2.0000  3.0000
       3.0  4.0000  5.0000  6.0000
      10.0  7.0000  8.0000  9.0000
   </tableData>

</table>
```

**Reply by bcoconni** [reply]
Just a reminder, `FGTurbine`  alters the thrust according to the throttle position `fcs/throttle-pos-norm`.
https://github.com/JSBSim-Team/jsbsim/blob/c68ef9f26f55ed9920f66bd12a9b2fb2ed40ef38/src/models/propulsion/FGTurbine.cpp#L211-L216
You'll also note that the actual thrust is the sum of the idle thrust and of the milthrust. So you need to either:
* Set `fcs/throttle-pos-norm` to 0.0 and `milthrust`to 0.0 or,
* Set `fcs/throttle-pos-norm` to 1.0 and `idlethrust`to 0.0

**Comment by seanmcleod** [comment]
Or what about using `<interpolate1d>`?

From an example @djlinse posted on the mailing list in 2013.

If you replaced `aero/qbar-psf` with say your own `fcs/engine-rating-value` and there are only discrete values then there would be no interpolation between them, it would effectively use it as an index.

```xml
<function name="bigWhatever1">
  <interpolate1d>
    <p> aero/qbar-psf </p>
    <v> 0 </v>  <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ...
                  <v> 0.80 </v>  <table> ... table 2 definition ...
                  <v> 0.90 </v>  <table> ... table 3 definition ...
                 </interpolate1d>
    <v> 65 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ...
                  <v> 0.80 </v>  <table> ... table 2 definition ...
                  <v> 0.90 </v>  <table> ... table 3 definition ...
                </interpolate1d>
    <v> 90 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ...
                  <v> 0.80 </v>  <table> ... table 2 definition ...
                  <v> 0.90 </v>  <table> ... table 3 definition ...
                </interpolate1d>
  </interpolate1d>
</function>
```

**Reply by agodemar** [reply]
@seanmcleod How would the config file overall structure look like? I mean, including variable tsfc (table lookup) and variable available thrust (table lookup), both dependent on engine rating, altitude, and Mach number.

**Comment by seanmcleod** [comment]
Actually here is @jonsberndt's originally posting to the mailing list (I've just added the xml formatting) explaining how to use `interpolate1d` to handle high dimensional data.

-----Original Message----- 
From: Jon S. Berndt
Sent: Monday, June 10, 2013 3:53 AM
To: 'Development issues'
Subject: [Jsbsim-devel] Multi-dimensional tables

It has long been the case that 3-dimensional tables are the limit of what we
can do in JSBSim. Over the past few weeks I have become aware of a way to do
n-dimensional tables, with additional capabilities, too.

Some lookup tables in simulation - particularly for aerodynamic data - can
be four, five, six, or even more dimensional. Not too long ago, an
interpolation function was added to JSBSim. It is called Interpolate1d.
Interpolate1d returns the result from a 1-dimensional interpolation of the
supplied values, with the value of the first immediate child element
representing the lookup value into the table, and the following pairs of
values representing the independent and dependent values. The first provided
child element is expected to be a property. The interpolation does not
extrapolate, but holds the highest value if the provided lookup value goes
outside of the provided range. The format is as follows:

```xml
<interpolate1d>
  {property, value, table, function}
  {property, value, table, function} {property, value, table, function}
  ...
</interpolate1d>
```

Example: If mach is 0.4, the interpolation will return 0.375. If mach is 1.5, the interpolation will return 0.60.

```xml
<interpolate1d>
  <p> velocities/mach </p>
  <v> 0.00 </v>  <v> 0.25 </v>
  <v> 0.80 </v>  <v> 0.50 </v>
  <v> 0.90 </v>  <v> 0.60 </v>
</interpolate1d>
```

The above example is very simplistic. A more involved example would use a
function in any argument (except the first). That means that the breakpoint
vector can be variable - which would probably not be common - but more
importantly the values in the lookup vector (second column) could be
function table elements of 1, 2, or 3 dimensions. The arguments could even
be nested interpolate1d elements. For example:

```xml
<function name="whatever">
  <interpolate1d>
    <p> velocities/mach </p>
    <v> 0.00 </v>  <table> ... table definition ... </table>
    <v> 0.80 </v>  <table> ... table definition ... </table>
    <v> 0.90 </v>  <table> ... table definition ... </table>
  </interpolate1d>
</function>
```

Carrying this further:

```xml
<function name="bigWhatever1">
  <interpolate1d>
    <p> aero/qbar-psf </p>
    <v> 0 </v>  <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 65 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 90 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
  </interpolate1d>
</function>
```

The above effectively gives a five dimensional lookup table. It would be big
and messy, in practice, but there it is. :-)

There is more, though. For very, very large aero databases there might be
times when some aero coefficients do not need to be calculated. For
instance, ground effects aero coefficients only need to be calculated close
to the ground. Why waste CPU cycles when the ground effects do not
contribute to the aero forces and moments? We can employ the `ifthen` element
to bypass expensive computations. The `ifthen` element works as follows:

If the value of the first immediate child element is 1, then the value of
the second immediate child element is returned, otherwise the value of the
third child element is returned.

```xml
<ifthen>
  {property, value, table, or other function element}
  {property, value, table, or other function element}
  {property, value, table, or other function element}
</ifthen>
```
Example: if flight-mode is greater than 2, then a value of 0.00 is returned, otherwise the value of the property `control/pitch-lag` is returned.

```xml
<ifthen>
  <gt> <p> executive/flight-mode </p> <v> 2 </v> </gt>
  <v> 0.00 </v>
  <p> control/pitch-lag </p>
</ifthen>
```

In our case, we could write the 5-dimensional table lookup as follows, returning a zero unless the gear is down:

```xml
<function name="propertyname">
  <ifthen>
    <lt> <p> position/altitudeMSL </p> <v> 90 </v> </lt>
    <interpolate1d>
      <p> aero/qbar-psf </p>
      <v> 0 </v>  <interpolate1d>
                    <p> velocities/mach </p>
                    <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                    <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                    <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                  </interpolate1d>
      <v> 65 </v> <interpolate1d>
                    <p> velocities/mach </p>
                    <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                    <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                    <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                  </interpolate1d>
      <v> 90 </v> <interpolate1d>
                    <p> velocities/mach </p>
                    <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                    <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                    <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                  </interpolate1d>
    </interpolate1d>
    <v> 0 </v>
  </ifthen>
</function>
```

The above example is non-sensical in a way, but the format is correct. I
have used this approach and it works well. Performance-wise it is good
because the tables do not get executed unless absolutely needed in the
lookup.

Jon

**Comment by agodemar** [comment]
@seanmcleod How the file `CFM56.xml` would be modified then?

**Reply by agodemar** [reply]
`<turbine_engine name="CFM56">`
`   <milthrust> 20000.0 </milthrust>`
`   <bypassratio>     5.9 </bypassratio> `
`   <tsfc>            ??? </tsfc> `
`   <function name="???"> `
`    <table> `
` `
`    </table> `
`   </function> `
`</turbine_engine>`

**Comment by seanmcleod** [comment]
So your `MilThrust` function would look like this.

So the 1st dimension is indexed on `fcs/engine-rating-value` then the next dimension is `velocities/mach` and then the `<table>` is a simple 1D lookup table based on `altitude` with the table values being the thrust value. So actually in this case you could skip the 2nd `interpolate1d` and just use a 2D table indexed with `mach` and `altitude`.

Although you had previously also mentioned having another data dimension, i.e. `temp-delta-ISA`.

```xml
<function name="MilThrust">
  <interpolate1d>
    <p> fcs/engine-rating-value </p>
    <v> 1 </v>  <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 2 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 3 </v> <interpolate1d>
                  <p> velocities/mach </p>
                  <v> 0.00 </v>  <table> ... table 1 definition ... </table>
                  <v> 0.80 </v>  <table> ... table 2 definition ... </table>
                  <v> 0.90 </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
  </interpolate1d>
</function>
```

Then you would have the equivalent for `tsfc`, copying and just changing `<function name="MilThrust>` to `<tsfc>` and obviously filling in the tsfc values in the tables as opposed to thrust values.

**Comment by seanmcleod** [comment]
So assuming the dimensions, `engine-rating`, `temp-delta-ISA`, `mach`, `altitude` here is one option.

So it first indexes based on the discrete `engine-rating` and then there are tables for each `temp-delta-ISA` that you have for each engine rating, and it will interpolate between these temperature differences after looking up the value via a 2D table based on `mach` and `altitude`.

```xml
<function name="MilThrust">
  <interpolate1d>
    <p> fcs/engine-rating-value </p>
    <v> 1 </v>  <interpolate1d>
                  <p> temp-delta-ISA </p>
                  <v> -10 </v>  
                            <table> 
                                 <independentVar lookup="row">velocities/mach</independentVar>
                                 <independentVar lookup="column">atmosphere/density-altitude</independentVar>
                                 <tableData>
                                        -10000       0   10000   20000   30000   40000   50000   60000
                                  0.0   1.2600  1.0000  0.7400  0.5340  0.3720  0.2410  0.1490  0.0
                                  0.2   1.1710  0.9340  0.6970  0.5060  0.3550  0.2310  0.1430  0.0
                                  0.4   1.1500  0.9210  0.6920  0.5060  0.3570  0.2330  0.1450  0.0
                                  0.6   1.1810  0.9510  0.7210  0.5320  0.3780  0.2480  0.1540  0.0
                                  0.8   1.2580  1.0200  0.7820  0.5820  0.4170  0.2750  0.1700  0.0
                                  1.0   1.3690  1.1200  0.8710  0.6510  0.4750  0.3150  0.1950  0.0
                                  1.2   0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0
                                  </tableData>
                           </table>
                  <v> 0   </v>  <table> ... table 2 definition ... </table>
                  <v> 10  </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 2 </v> <interpolate1d>
                  <p> temp-delta-ISA </p>
                  <v> -10 </v>  <table> ... table 1 definition ... </table>
                  <v> 0   </v>  <table> ... table 2 definition ... </table>
                  <v> 10  </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
    <v> 3 </v> <interpolate1d>
                  <p> temp-delta-ISA </p>
                  <v> -10 </v>  <table> ... table 1 definition ... </table>
                  <v> 0   </v>  <table> ... table 2 definition ... </table>
                  <v> 10  </v>  <table> ... table 3 definition ... </table>
                </interpolate1d>
  </interpolate1d>
</function>
```
